java/ql/src/experimental/Security/CWE/CWE-470/UnsafeReflection.ql

```diff
@@ -0,0 +1,49 @@
+/**
+ * @name Use of externally-controlled input to select classes or code ('unsafe reflection')
+ * @description Use external input with reflection function to select the class or code to
+ *              be used, which brings serious security risks.
+ * @kind path-problem
+ * @problem.severity error
+ * @precision high
+ * @id java/unsafe-reflection
+ * @tags security
+ *       external/cwe/cwe-470
+ */
+
+import java
+import UnsafeReflectionLib
+import semmle.code.java.dataflow.FlowSources
+import DataFlow::PathGraph
+
+class ContainsSanitizer extends DataFlow::BarrierGuard {
+  ContainsSanitizer() { this.(MethodAccess).getMethod().hasName("contains") }
+
+  override predicate checks(Expr e, boolean branch) {
+    e = this.(MethodAccess).getArgument(0) and branch = false
+  }
+}
+
+class EqualsSanitizer extends DataFlow::BarrierGuard {
+  EqualsSanitizer() { this.(MethodAccess).getMethod().hasName("equals") }
+
+  override predicate checks(Expr e, boolean branch) {
+    e = [this.(MethodAccess).getArgument(0), this.(MethodAccess).getQualifier()] and
```

Comment:
Is line 30 a new usage?

Reply:
https://codeql.github.com/docs/ql-language-reference/expressions/#set-literal-expressions

Reply:
thanks